File: bindings/java/src/genlang/Document.java
package genlang;

import com.sun.jna.Memory;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Document implements AutoCloseable {
    private Pointer ctx;
    private Pointer doc;

    private Document(Pointer ctx, Pointer doc) {
        this.ctx = ctx;
        this.doc = doc;
    }

    public static String version() {
        return utf8(Lib.INSTANCE.gen_version());
    }

    public static Document parse(String source) {
        return open(source, false);
    }

    public static Document load(String path) {
        return open(path, true);
    }

    private static Document open(String arg, boolean loadFile) {
        Pointer ctx = Lib.INSTANCE.gen_context_create();
        if (ctx == null) {
            throw new GenLangException(2, "out of memory");
        }
        Memory cArg = utf8z(arg);
        PointerByReference out = new PointerByReference();
        int rc = loadFile
                ? Lib.INSTANCE.gen_document_load_file(ctx, cArg, out)
                : Lib.INSTANCE.gen_document_parse(ctx, cArg, out);
        if (rc != 0) {
            Pointer err = Lib.INSTANCE.gen_context_last_error(ctx);
            GenLangException ex = new GenLangException(
                    rc,
                    utf8(Lib.INSTANCE.gen_error_message(err)),
                    (int) Lib.INSTANCE.gen_error_line(err).longValue(),
                    (int) Lib.INSTANCE.gen_error_column(err).longValue(),
                    utf8(Lib.INSTANCE.gen_error_path(err))
            );
            Lib.INSTANCE.gen_context_free(ctx);
            throw ex;
        }
        return new Document(ctx, out.getValue());
    }

    @Override
    public void close() {
        if (doc != null) {
            Lib.INSTANCE.gen_document_free(doc);
            doc = null;
        }
        if (ctx != null) {
            Lib.INSTANCE.gen_context_free(ctx);
            ctx = null;
        }
    }

    private void ensureOpen() {
        if (doc == null) {
            throw new IllegalStateException("document is closed");
        }
    }

    public List<String> types() {
        ensureOpen();
        int n = (int) Lib.INSTANCE.gen_type_count(doc).longValue();
        List<String> names = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            PointerByReference name = new PointerByReference();
            Lib.INSTANCE.gen_type_name(doc, new Lib.SizeT(i), name);
            names.add(utf8(name.getValue()));
        }
        return names;
    }

    public List<String> entities() {
        ensureOpen();
        int n = (int) Lib.INSTANCE.gen_entity_count(doc).longValue();
        List<String> names = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            PointerByReference name = new PointerByReference();
            Lib.INSTANCE.gen_entity_name(doc, new Lib.SizeT(i), name);
            names.add(utf8(name.getValue()));
        }
        return names;
    }

    public String typeParent(String name) {
        ensureOpen();
        PointerByReference parent = new PointerByReference();
        int rc = Lib.INSTANCE.gen_type_parent(doc, utf8z(name), parent);
        if (rc != 0) {
            throw new GenLangException(rc, "unknown type '" + name + "'");
        }
        return utf8(parent.getValue());
    }

    public String entityType(String name) {
        ensureOpen();
        PointerByReference type = new PointerByReference();
        int rc = Lib.INSTANCE.gen_entity_type_name(doc, utf8z(name), type);
        if (rc != 0) {
            throw new GenLangException(rc, "unknown entity '" + name + "'");
        }
        return utf8(type.getValue());
    }

    public Object entity(String name) {
        ensureOpen();
        PointerByReference value = new PointerByReference();
        int rc = Lib.INSTANCE.gen_entity_value(doc, utf8z(name), value);
        if (rc != 0) {
            throw new GenLangException(rc, "unknown entity '" + name + "'");
        }
        return valueFromNative(value.getValue());
    }

    public Object get(String path) {
        ensureOpen();
        PointerByReference value = new PointerByReference();
        int rc = Lib.INSTANCE.gen_get(doc, utf8z(path), value);
        if (rc != 0) {
            throw new GenLangException(rc, "cannot evaluate path '" + path + "'");
        }
        try {
            return valueFromNative(value.getValue());
        } finally {
            Lib.INSTANCE.gen_value_free(value.getValue());
        }
    }

    public List<String> ancestors(String name) {
        ensureOpen();
        PointerByReference result = new PointerByReference();
        int rc = Lib.INSTANCE.gen_ancestors_of(doc, utf8z(name), result);
        return namesFromQuery(rc, result.getValue(), "unknown type '" + name + "'");
    }

    public List<String> members(String setName) {
        ensureOpen();
        PointerByReference result = new PointerByReference();
        int rc = Lib.INSTANCE.gen_set_members(doc, utf8z(setName), result);
        return namesFromQuery(rc, result.getValue(), "unknown set '" + setName + "'");
    }

    public boolean isMember(String setName, String entity) {
        ensureOpen();
        IntByReference flag = new IntByReference();
        int rc = Lib.INSTANCE.gen_is_member(doc, utf8z(setName), utf8z(entity), flag);
        if (rc != 0) {
            throw new GenLangException(rc, "unknown set '" + setName + "'");
        }
        return flag.getValue() != 0;
    }

    public String serialize() {
        ensureOpen();
        PointerByReference text = new PointerByReference();
        int rc = Lib.INSTANCE.gen_document_serialize(doc, text, null);
        if (rc != 0) {
            throw new GenLangException(rc, "serialize failed");
        }
        try {
            return utf8(text.getValue());
        } finally {
            Lib.INSTANCE.gen_string_free(text.getValue());
        }
    }

    public String toJson() {
        ensureOpen();
        PointerByReference text = new PointerByReference();
        int rc = Lib.INSTANCE.gen_document_to_json(doc, text, null);
        if (rc != 0) {
            throw new GenLangException(rc, "json conversion failed");
        }
        try {
            return utf8(text.getValue());
        } finally {
            Lib.INSTANCE.gen_string_free(text.getValue());
        }
    }

    public String toYaml() {
        ensureOpen();
        PointerByReference text = new PointerByReference();
        int rc = Lib.INSTANCE.gen_document_to_yaml(doc, text, null);
        if (rc != 0) {
            throw new GenLangException(rc, "yaml conversion failed");
        }
        try {
            return utf8(text.getValue());
        } finally {
            Lib.INSTANCE.gen_string_free(text.getValue());
        }
    }

    private static List<String> namesFromQuery(int rc, Pointer result, String error) {
        if (rc != 0) {
            throw new GenLangException(rc, error);
        }
        int n = (int) Lib.INSTANCE.gen_query_result_count(result).longValue();
        List<String> names = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            names.add(utf8(Lib.INSTANCE.gen_query_result_name(result, new Lib.SizeT(i))));
        }
        Lib.INSTANCE.gen_query_result_free(result);
        return names;
    }

    private static Object valueFromNative(Pointer value) {
        if (value == null) {
            return null;
        }
        switch (Lib.INSTANCE.gen_value_type(value)) {
            case 0:
                return null;
            case 1:
                return Lib.INSTANCE.gen_value_bool(value) != 0;
            case 2:
                return Lib.INSTANCE.gen_value_int(value);
            case 3:
                return Lib.INSTANCE.gen_value_float(value);
            case 4:
                return utf8(Lib.INSTANCE.gen_value_string(value));
            case 5: {
                int n = (int) Lib.INSTANCE.gen_value_list_count(value).longValue();
                List<Object> list = new ArrayList<>(n);
                for (int i = 0; i < n; i++) {
                    PointerByReference item = new PointerByReference();
                    Lib.INSTANCE.gen_value_list_get(value, new Lib.SizeT(i), item);
                    list.add(valueFromNative(item.getValue()));
                }
                return list;
            }
            case 6: {
                int n = (int) Lib.INSTANCE.gen_value_object_count(value).longValue();
                Map<String, Object> obj = new LinkedHashMap<>();
                for (int i = 0; i < n; i++) {
                    PointerByReference key = new PointerByReference();
                    PointerByReference item = new PointerByReference();
                    Lib.INSTANCE.gen_value_object_key(value, new Lib.SizeT(i), key);
                    Lib.INSTANCE.gen_value_object_get_index(value, new Lib.SizeT(i), item);
                    obj.put(utf8(key.getValue()), valueFromNative(item.getValue()));
                }
                return obj;
            }
            case 7:
                return new Ref(utf8(Lib.INSTANCE.gen_value_reference(value)));
            default:
                return null;
        }
    }

    private static Memory utf8z(String s) {
        byte[] bytes = s.getBytes(StandardCharsets.UTF_8);
        Memory mem = new Memory(bytes.length + 1L);
        mem.write(0, bytes, 0, bytes.length);
        mem.setByte(bytes.length, (byte) 0);
        return mem;
    }

    private static String utf8(Pointer p) {
        if (p == null) {
            return null;
        }
        return p.getString(0, "UTF-8");
    }
}

File: bindings/csharp/GenLang/Document.cs
using System.Runtime.InteropServices;

namespace GenLang;

public sealed class Ref
{
    public Ref(string name) => Name = name;
    public string Name { get; }
    public override string ToString() => "@" + Name;
}

public sealed class GenLangException : Exception
{
    public GenLangException(int code, string message, int line = 0, int column = 0, string? path = null)
        : base(line > 0 ? $"{path ?? "<input>"}:{line}:{column}: {message}" : message)
    {
        Code = code;
        Line = line;
        Column = column;
        Path = path;
    }

    public int Code { get; }
    public int Line { get; }
    public int Column { get; }
    public string? Path { get; }
}

public sealed class Document : IDisposable
{
    private IntPtr _ctx;
    private IntPtr _doc;

    private Document(IntPtr ctx, IntPtr doc)
    {
        _ctx = ctx;
        _doc = doc;
    }

    public static string Version() => Native.Utf8(Native.gen_version()) ?? "";

    public static Document Parse(string source) => Open(source, loadFile: false);

    public static Document Load(string path) => Open(path, loadFile: true);

    private static Document Open(string arg, bool loadFile)
    {
        var ctx = Native.gen_context_create();
        if (ctx == IntPtr.Zero)
        {
            throw new GenLangException(2, "out of memory");
        }
        var cArg = Native.Utf8Alloc(arg);
        try
        {
            var rc = loadFile
                ? Native.gen_document_load_file(ctx, cArg, out var doc)
                : Native.gen_document_parse(ctx, cArg, out doc);
            if (rc != 0)
            {
                var err = Native.gen_context_last_error(ctx);
                var ex = new GenLangException(
                    rc,
                    Native.Utf8(Native.gen_error_message(err)) ?? "error",
                    (int)Native.gen_error_line(err),
                    (int)Native.gen_error_column(err),
                    Native.Utf8(Native.gen_error_path(err))
                );
                Native.gen_context_free(ctx);
                throw ex;
            }
            return new Document(ctx, doc);
        }
        finally
        {
            Marshal.FreeHGlobal(cArg);
        }
    }

    public void Dispose()
    {
        if (_doc != IntPtr.Zero)
        {
            Native.gen_document_free(_doc);
            _doc = IntPtr.Zero;
        }
        if (_ctx != IntPtr.Zero)
        {
            Native.gen_context_free(_ctx);
            _ctx = IntPtr.Zero;
        }
        GC.SuppressFinalize(this);
    }

    ~Document() => Dispose();

    private void EnsureOpen()
    {
        if (_doc == IntPtr.Zero)
        {
            throw new ObjectDisposedException(nameof(Document));
        }
    }

    public IReadOnlyList<string> Types
    {
        get
        {
            EnsureOpen();
            var n = (int)Native.gen_type_count(_doc);
            var names = new List<string>(n);
            for (var i = 0; i < n; i++)
            {
                Native.gen_type_name(_doc, (UIntPtr)i, out var name);
                names.Add(Native.Utf8(name) ?? "");
            }
            return names;
        }
    }

    public IReadOnlyList<string> Entities
    {
        get
        {
            EnsureOpen();
            var n = (int)Native.gen_entity_count(_doc);
            var names = new List<string>(n);
            for (var i = 0; i < n; i++)
            {
                Native.gen_entity_name(_doc, (UIntPtr)i, out var name);
                names.Add(Native.Utf8(name) ?? "");
            }
            return names;
        }
    }

    public string? TypeParent(string name)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(name);
        try
        {
            var rc = Native.gen_type_parent(_doc, cName, out var parent);
            if (rc != 0)
            {
                throw new GenLangException(rc, $"unknown type '{name}'");
            }
            return Native.Utf8(parent);
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public string? EntityType(string name)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(name);
        try
        {
            var rc = Native.gen_entity_type_name(_doc, cName, out var type);
            if (rc != 0)
            {
                throw new GenLangException(rc, $"unknown entity '{name}'");
            }
            return Native.Utf8(type);
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public object? Entity(string name)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(name);
        try
        {
            var rc = Native.gen_entity_value(_doc, cName, out var value);
            if (rc != 0)
            {
                throw new GenLangException(rc, $"unknown entity '{name}'");
            }
            return ValueFromNative(value);
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public object? Get(string path)
    {
        EnsureOpen();
        var cPath = Native.Utf8Alloc(path);
        try
        {
            var rc = Native.gen_get(_doc, cPath, out var value);
            if (rc != 0)
            {
                throw new GenLangException(rc, $"cannot evaluate path '{path}'");
            }
            try
            {
                return ValueFromNative(value);
            }
            finally
            {
                Native.gen_value_free(value);
            }
        }
        finally
        {
            Marshal.FreeHGlobal(cPath);
        }
    }

    public IReadOnlyList<string> Ancestors(string name)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(name);
        try
        {
            var rc = Native.gen_ancestors_of(_doc, cName, out var result);
            return NamesFromQuery(rc, result, $"unknown type '{name}'");
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public IReadOnlyList<string> Members(string setName)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(setName);
        try
        {
            var rc = Native.gen_set_members(_doc, cName, out var result);
            return NamesFromQuery(rc, result, $"unknown set '{setName}'");
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public IReadOnlyList<string> EntitiesOf(string typeName)
    {
        EnsureOpen();
        var cName = Native.Utf8Alloc(typeName);
        try
        {
            var rc = Native.gen_entities_of(_doc, cName, out var result);
            return NamesFromQuery(rc, result, $"unknown type '{typeName}'");
        }
        finally
        {
            Marshal.FreeHGlobal(cName);
        }
    }

    public bool IsMember(string setName, string entity)
    {
        EnsureOpen();
        var cSet = Native.Utf8Alloc(setName);
        var cEnt = Native.Utf8Alloc(entity);
        try
        {
            var rc = Native.gen_is_member(_doc, cSet, cEnt, out var flag);
            if (rc != 0)
            {
                throw new GenLangException(rc, $"unknown set '{setName}'");
            }
            return flag != 0;
        }
        finally
        {
            Marshal.FreeHGlobal(cSet);
            Marshal.FreeHGlobal(cEnt);
        }
    }

    public string Serialize()
    {
        EnsureOpen();
        var rc = Native.gen_document_serialize(_doc, out var text, IntPtr.Zero);
        if (rc != 0)
        {
            throw new GenLangException(rc, "serialize failed");
        }
        try
        {
            return Native.Utf8(text) ?? "";
        }
        finally
        {
            Native.gen_string_free(text);
        }
    }

    public string ToJson()
    {
        EnsureOpen();
        var rc = Native.gen_document_to_json(_doc, out var text, IntPtr.Zero);
        if (rc != 0)
        {
            throw new GenLangException(rc, "json conversion failed");
        }
        try
        {
            return Native.Utf8(text) ?? "";
        }
        finally
        {
            Native.gen_string_free(text);
        }
    }

    public string ToYaml()
    {
        EnsureOpen();
        var rc = Native.gen_document_to_yaml(_doc, out var text, IntPtr.Zero);
        if (rc != 0)
        {
            throw new GenLangException(rc, "yaml conversion failed");
        }
        try
        {
            return Native.Utf8(text) ?? "";
        }
        finally
        {
            Native.gen_string_free(text);
        }
    }

    private static IReadOnlyList<string> NamesFromQuery(int rc, IntPtr result, string error)
    {
        if (rc != 0)
        {
            throw new GenLangException(rc, error);
        }
        var n = (int)Native.gen_query_result_count(result);
        var names = new List<string>(n);
        for (var i = 0; i < n; i++)
        {
            names.Add(Native.Utf8(Native.gen_query_result_name(result, (UIntPtr)i)) ?? "");
        }
        Native.gen_query_result_free(result);
        return names;
    }

    private static object? ValueFromNative(IntPtr value)
    {
        if (value == IntPtr.Zero)
        {
            return null;
        }
        switch (Native.gen_value_type(value))
        {
            case 0:
                return null;
            case 1:
                return Native.gen_value_bool(value) != 0;
            case 2:
                return Native.gen_value_int(value);
            case 3:
                return Native.gen_value_float(value);
            case 4:
                return Native.Utf8(Native.gen_value_string(value)) ?? "";
            case 5:
            {
                var n = (int)Native.gen_value_list_count(value);
                var list = new List<object?>(n);
                for (var i = 0; i < n; i++)
                {
                    Native.gen_value_list_get(value, (UIntPtr)i, out var item);
                    list.Add(ValueFromNative(item));
                }
                return list;
            }
            case 6:
            {
                var n = (int)Native.gen_value_object_count(value);
                var obj = new Dictionary<string, object?>(n);
                for (var i = 0; i < n; i++)
                {
                    Native.gen_value_object_key(value, (UIntPtr)i, out var key);
                    Native.gen_value_object_get_index(value, (UIntPtr)i, out var item);
                    obj[Native.Utf8(key) ?? ""] = ValueFromNative(item);
                }
                return obj;
            }
            case 7:
                return new Ref(Native.Utf8(Native.gen_value_reference(value)) ?? "");
            default:
                return null;
        }
    }
}
